Gems/ScriptCanvas/Code/Builder/ScriptCanvasBuilderDataSystem.cpp

```diff
@@ -88,69 +93,200 @@ namespace ScriptCanvasBuilder
+        if (auto iter = m_assets.find(sourceHandle.Id()); iter != m_assets.end())
+        {
+            return iter->second;
+        }
+        else
+        {
+            return MonitorAsset(sourceHandle.Id());
+        }
+    }
+
+    void DataSystem::OnAssetError(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        const auto assetIdGuid = asset.GetId().m_guid;
+        DATA_SYSTEM_STATUS
+            ( "ScriptCanvas"
+            , "DataSystem received OnAssetError: %s : %s"
+            , asset.GetHint().c_str()
+            , assetIdGuid.ToString<AZStd::string>().c_str());
+
+        auto& buildResult = m_assets[assetIdGuid];
+        buildResult.data = asset;
+        buildResult.status = BuilderAssetStatus::Error;
+        DataSystemAssetNotificationsBus::Event
+            ( assetIdGuid
+            , &DataSystemAssetNotifications::OnAssetNotReady);
+    }
+
+    void DataSystem::OnAssetReady(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        DATA_SYSTEM_STATUS
+            ( "ScriptCanvas"
+            , "DataSystem received OnAssetReady: %s : %s"
+            , asset.GetHint().c_str()
+            , asset.GetId().m_guid.ToString<AZStd::string>().c_str());
+        ReportReadyFilter(asset);     
+    }
+
+    void DataSystem::ReportReady(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        using namespace ScriptCanvas;
+        const auto assetIdGuid = asset.GetId().m_guid;
+
+        auto& buildResult = m_assets[assetIdGuid];
+        buildResult.data = asset;
+        buildResult.data.SetAutoLoadBehavior(AZ::Data::AssetLoadBehavior::PreLoad);
+        
+        if (auto result = IsPreloaded(buildResult.data); result != IsPreloadedResult::Yes)
+        {
+            AZ_Error("ScriptCanvas"
+                , false
+                , "DataSystem received ready for asset that was not loaded: %s-%s"
+                , buildResult.data.GetHint().c_str()
+                , buildResult.data.GetId().ToString<AZStd::string>().c_str()
+            );
+
+            buildResult.status = BuilderAssetStatus::Error;
+            DataSystemAssetNotificationsBus::Event
+                ( assetIdGuid
+                , &DataSystemAssetNotifications::OnAssetNotReady);
+        }
+        else
+        {
+            buildResult.status = BuilderAssetStatus::Ready;
+            DataSystemAssetNotificationsBus::Event
+                ( assetIdGuid
+                , &DataSystemAssetNotifications::OnReady
+                , buildResult.data);
+        }
+    }
+
+    void DataSystem::ReportReadyFilter(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        using namespace ScriptCanvas;
+
+        SCRIPT_SYSTEM_SCRIPT_STATUS("ScriptCanvas", "DataSystem::ReportReadyFilter received a runtime asset, queuing Lua script processing.");
+        AZ::SystemTickBus::QueueFunction([this, asset]()
+        {
+            SCRIPT_SYSTEM_SCRIPT_STATUS("ScriptCanvas", "DataSystem::ReportReadyFilter executing Lua script processing.");
```

Comment:
Normally, the Data system would be able to immediately broadcast that the asset is ready. However there is a problem with holding stale Lua assets in memory, and a fresh copy must be loaded. This is a known issue with a fix coming soon.